### ISSUE TITLE
(Somewhat) silly question: what is a 'dataframe'?

### ISSUE THREAD

**Issue by kenny-evitt** [elixir-nx/explorer#80]
## TLDR

I'm just curious 🙂

## Too long to read

I *think* I know what 'dataframe' means. I've been thinking of writing my own 'datacube' Elixir library, basically an 'in-memory database table'.

But I'm still not exactly sure what 'dataframe' means and some other people are similarly (somewhat) confused; see:

- [Explorer - Dataframes for Elixir : elixir](https://www.reddit.com/r/elixir/comments/qn0632/explorer_dataframes_for_elixir/)

What's a little annoying/frustrating (in a very specific kind of way that these things sometimes are nowadays) is that doing a web (Google) search for "dataframe" or "dataframe definition" or "dataframe wikipedia" all return results about specific dataframe products/projects and all of them seem to assume you already know what a 'dataframe' is.

Here's an example result that's not quite entirely satisfying:

- [What are DataFrames? - Databricks](https://databricks.com/glossary/what-are-dataframes)

It kinda seems like 'dataframe' is so abstract that regular RDMS DB *tables* themselves are, in a sense, 'dataframes'. But that doesn't seem quite right to me!

The docs are a little helpful:

> The `Explorer` library is a set of functions and data types to work with tabular data in Elixir.

In what sense is `Ecto` also (or not) a 'dataframe' library for Elixir?

**Comment by kenny-evitt** [comment]
This post (on a blog about a new Unix shell, of all things) actually seems to be trying to answer this very question:

- [What Is a Data Frame? (In Python, R, and SQL)](https://www.oilshell.org/blog/2018/11/30.html)

From that post:

> # What Is a Data Frame? (In Python, R, and SQL)

> A data frame is a **table-like data structure** available in languages like R and Python. Statisticians, scientists, and programmers use them in data analysis code.
>
> ...
>
> 1. In the [last post](https://www.oilshell.org/blog/2018/11/15.html), I analyzed Oil's code [with data frames](https://www.oilshell.org/blog/2018/11/15.html#toc_12).
> 
>    When writing the praise in that section, I looked for a friendly intro to the idea, but the best hit was from an [inactive Clojure project](https://github.com/mobileink/data.frame/wiki/What-is-a-Data-Frame%3F). Surprisingly, Wikipedia doesn't have a page on data frames! This post aims to fill that gap.
>
> ...
>
> ## Typical Operations
> 
> Before I show source code, here's some more color on how data frames are used. You may:
> 
> - **Create** a data frame from the results of a SQL query, or from a CSV file. The columns have **types** like string, number, and date.
> - **Filter** a data frame down to the rows and columns of interest.
> - **Clean** its values with arithmetic and string operations.
> - **Summarize** groups of rows.
> - **Compute** new columns based on existing columns.
> - **Join** a data frame with others, for further analysis.
> - **Plot** one column vs. another ([in many different ways](https://ggplot2.tidyverse.org/reference/#section-layer-geoms))
> - Mathematically **model** one column as a function of another — e.g. with linear regression.
> 
> Logically, data frames look like SQL tables, but:
> 
> - They're manipulated directly in the memory of your program. They don't live on a remote server.
> - They more suited to complex, "offline" analysis with statistics, visualization, and mathematical modeling.

The first item in that last list doesn't seem to be entirely true – or at least not according to Databricks about their *Spark DataFrames*:

> ... a Spark DataFrame can span thousands of computers

So SQL tables are *kinda* 'dataframes' but *typically* a 'dataframe' 'lives' closer to the 'processing', e.g. 'in-memory' versus accessible over the network or (for Elixir) outside the BEAM VM (?).

**Comment by josevalim** [comment]
Good point. As things evolve, should we make sure to write the docs in a way we don't assume everyone has a ML/Pandas background.
